File: LintCode-Java/ReverseLinkedListII.java
/**
 * Definition for ListNode
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class ReverseLinkedListII {
    /**
     * @param ListNode head is the head of the linked list
     * @oaram m and n
     * @return: The head of the reversed ListNode
     */

    // V1, O(n)
    public ListNode reverseBetween(ListNode head, int m , int n) {
        if (head == null || m >= n) {
            return head;
        }

        ListNode dummy = new ListNode(0);
        dummy.next = head;
        head = dummy;

        for (int i = 1; i < m; i++) {
            if (head == null) {
                return null;
            }
            head = head.next;
        }

        ListNode premNode = head;
        ListNode mNode = head.next;

        ListNode nNode = mNode;
        ListNode postnNode = mNode.next;

        for (int i = m; i < n; i++) {
            if (postnNode == null) {
                return null;
            }
            ListNode tmp = postnNode.next;
            postnNode.next = nNode;
            nNode = postnNode;
            postnNode = tmp;
        }

        premNode.next = nNode;
        mNode.next = postnNode;

        return dummy.next;

    }
}

File: LeetCode/python/reverse_linked_list_II.py
# Reverse a linked list from position m to n. Do it in-place and in one-pass.

# For example:
# Given 1->2->3->4->5->NULL, m = 2 and n = 4,

# return 1->4->3->2->5->NULL.

# Note:
# Given m, n satisfy the following condition:
# 1 ≤ m ≤ n ≤ length of list.

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param m, an integer
    # @param n, an integer
    # @return a ListNode
    def reverseBetween_1(self, head, m, n):
        if head == None: return
        helper = ListNode(0); helper.next = head
        prev = helper
        for i in xrange(1, m):
            prev = prev.next
        # start: m.prev; end: m
        start = prev
        end = prev.next
        # prev: m; curr: m.next
        prev, curr = prev.next, prev.next.next
        for i in xrange(m, n):
            tmp = curr.next
            curr.next = prev
            prev = curr
            curr = tmp
        # prev: n; curr: n.next
        start.next = prev
        end.next = curr
        return helper.next
                